File: infer.py
import argparse
import torch

from PIL import Image
from torchvision import transforms

from model import Model

parser = argparse.ArgumentParser()
parser.add_argument('-c', '--checkpoint', type=str, required=True, help='path to checkpoint, e.g. ./logs/model-100.pth')
parser.add_argument('input', type=str, help='path to input image')


def _infer(path_to_checkpoint_file, path_to_input_image):
    model = Model()
    model.restore(path_to_checkpoint_file)
    model.cuda()

    with torch.no_grad():
        transform = transforms.Compose([
            transforms.Resize([64, 64]),
            transforms.CenterCrop([54, 54]),
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
        ])

        image = Image.open(path_to_input_image)
        image = image.convert('RGB')
        image = transform(image)
        images = image.unsqueeze(dim=0).cuda()

        length_logits, digit1_logits, digit2_logits, digit3_logits, digit4_logits, digit5_logits = model.eval()(images)

        length_prediction = length_logits.max(1)[1]
        digit1_prediction = digit1_logits.max(1)[1]
        digit2_prediction = digit2_logits.max(1)[1]
        digit3_prediction = digit3_logits.max(1)[1]
        digit4_prediction = digit4_logits.max(1)[1]
        digit5_prediction = digit5_logits.max(1)[1]

        print('length:', length_prediction.item())
        print('digits:', digit1_prediction.item(), digit2_prediction.item(), digit3_prediction.item(), digit4_prediction.item(), digit5_prediction.item())


def main(args):
    path_to_checkpoint_file = args.checkpoint
    path_to_input_image = args.input

    _infer(path_to_checkpoint_file, path_to_input_image)


if __name__ == '__main__':
    main(parser.parse_args())

File: cpp/infer.cpp
#include <torch/script.h>
#include <torch/utils.h>

#include <opencv2/opencv.hpp>


void _infer(const std::string& pathToCheckpointFile, const std::string& pathToInputImage) {
    std::shared_ptr<torch::jit::script::Module> model = torch::jit::load(pathToCheckpointFile);
    model->to(torch::kCUDA);

    torch::NoGradGuard noGradGuard;

    cv::Mat image;
    image = cv::imread(pathToInputImage, CV_LOAD_IMAGE_COLOR);

    cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
    cv::resize(image, image, cv::Size(64, 64));
    image = image(cv::Rect((64 - 54) / 2, (64 - 54) / 2, 54, 54)).clone();  // it's necessary to clone the ROI image to make elements continuous

    torch::Tensor images_tensor = torch::from_blob(image.data, {image.rows, image.cols, image.channels()}, torch::kByte).unsqueeze(0);
    images_tensor = images_tensor.permute({0, 3, 1, 2});  // BHWC to BCHW
    images_tensor = images_tensor.toType(torch::kFloat);
    images_tensor = images_tensor.div(255);
    images_tensor = images_tensor.sub(0.5).div(0.5);
    images_tensor = images_tensor.to(torch::kCUDA);

    auto outputs = model->forward({images_tensor}).toTuple()->elements();
    torch::Tensor length_logits = outputs[0].toTensor();
    torch::Tensor digit1_logits = outputs[1].toTensor();
    torch::Tensor digit2_logits = outputs[2].toTensor();
    torch::Tensor digit3_logits = outputs[3].toTensor();
    torch::Tensor digit4_logits = outputs[4].toTensor();
    torch::Tensor digit5_logits = outputs[5].toTensor();

    auto length_prediction = std::get<1>(length_logits.max(1)).item<int>();
    auto digit1_prediction = std::get<1>(digit1_logits.max(1)).item<int>();
    auto digit2_prediction = std::get<1>(digit2_logits.max(1)).item<int>();
    auto digit3_prediction = std::get<1>(digit3_logits.max(1)).item<int>();
    auto digit4_prediction = std::get<1>(digit4_logits.max(1)).item<int>();
    auto digit5_prediction = std::get<1>(digit5_logits.max(1)).item<int>();

    printf("length: %d \n", length_prediction);
    printf("digits: %d %d %d %d %d \n", digit1_prediction, digit2_prediction, digit3_prediction, digit4_prediction, digit5_prediction);
}

int main(int argc, const char* argv[]) {
    if (argc != 3) {
        std::cerr << "usage: infer <path-to-checkpoint-file> <path-to-input-image> \n";
        return -1;
    }

    const std::string pathToCheckpointFile = argv[1];
    const std::string pathToInputImage = argv[2];

    _infer(pathToCheckpointFile, pathToInputImage);

    return 0;
}